{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4fb02930",
   "metadata": {},
   "source": [
    "To perform the EDA we need to type several lines of code. \n",
    "Here auto-visualization library comes into the play, \n",
    "which can perform all these tasks using just 1 line of code. \n",
    "Some of these auto-visualization tools are:\n",
    "\n",
    "1. Pandas-Profiling\n",
    "2. Sweetviz\n",
    "3. Autoviz\n",
    "4. D-Tale"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e2c5670",
   "metadata": {},
   "source": [
    "### In this Notebook will discuss about D-Tale"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c9acd8fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting dtale\n",
      "  Downloading dtale-2.8.1-py2.py3-none-any.whl (12.8 MB)\n",
      "Collecting lz4\n",
      "  Downloading lz4-4.0.2-cp39-cp39-win_amd64.whl (98 kB)\n",
      "Collecting strsimpy\n",
      "  Downloading strsimpy-0.2.1-py3-none-any.whl (45 kB)\n",
      "Requirement already satisfied: numpy in c:\\users\\use\\anaconda3\\lib\\site-packages (from dtale) (1.21.5)\n",
      "Collecting dash-bootstrap-components\n",
      "  Downloading dash_bootstrap_components-1.2.1-py3-none-any.whl (216 kB)\n",
      "Requirement already satisfied: xarray in c:\\users\\use\\anaconda3\\lib\\site-packages (from dtale) (0.20.1)\n",
      "Requirement already satisfied: requests in c:\\users\\use\\anaconda3\\lib\\site-packages (from dtale) (2.27.1)\n",
      "Requirement already satisfied: matplotlib in c:\\users\\use\\anaconda3\\lib\\site-packages (from dtale) (3.5.1)\n",
      "Requirement already satisfied: pandas in c:\\users\\use\\anaconda3\\lib\\site-packages (from dtale) (1.4.2)\n",
      "Collecting kaleido\n",
      "  Downloading kaleido-0.2.1-py2.py3-none-win_amd64.whl (65.9 MB)\n",
      "Requirement already satisfied: six in c:\\users\\use\\anaconda3\\lib\\site-packages (from dtale) (1.16.0)\n",
      "Requirement already satisfied: future>=0.14.0 in c:\\users\\use\\anaconda3\\lib\\site-packages (from dtale) (0.18.2)\n",
      "Requirement already satisfied: itsdangerous in c:\\users\\use\\anaconda3\\lib\\site-packages (from dtale) (2.0.1)\n",
      "Requirement already satisfied: seaborn in c:\\users\\use\\anaconda3\\lib\\site-packages (from dtale) (0.11.2)\n",
      "Requirement already satisfied: scipy in c:\\users\\use\\anaconda3\\lib\\site-packages (from dtale) (1.7.3)\n",
      "Collecting dash>=2.0.0\n",
      "  Downloading dash-2.6.2-py3-none-any.whl (9.8 MB)\n",
      "Requirement already satisfied: et-xmlfile in c:\\users\\use\\anaconda3\\lib\\site-packages (from dtale) (1.1.0)\n",
      "Requirement already satisfied: certifi in c:\\users\\use\\anaconda3\\lib\\site-packages (from dtale) (2021.10.8)\n",
      "Requirement already satisfied: scikit-learn in c:\\users\\use\\anaconda3\\lib\\site-packages (from dtale) (1.0.2)\n",
      "Requirement already satisfied: plotly>=5.0.0 in c:\\users\\use\\anaconda3\\lib\\site-packages (from dtale) (5.6.0)\n",
      "Collecting flask-ngrok\n",
      "  Downloading flask_ngrok-0.0.25-py3-none-any.whl (3.1 kB)\n",
      "Collecting squarify\n",
      "  Downloading squarify-0.4.3-py3-none-any.whl (4.3 kB)\n",
      "Requirement already satisfied: Flask in c:\\users\\use\\anaconda3\\lib\\site-packages (from dtale) (1.1.2)\n",
      "Collecting dash-daq\n",
      "  Downloading dash_daq-0.5.0.tar.gz (642 kB)\n",
      "Collecting Flask-Compress\n",
      "  Downloading Flask_Compress-1.13-py3-none-any.whl (7.9 kB)\n",
      "Requirement already satisfied: statsmodels in c:\\users\\use\\anaconda3\\lib\\site-packages (from dtale) (0.13.2)\n",
      "Collecting dash-colorscales\n",
      "  Downloading dash_colorscales-0.0.4.tar.gz (62 kB)\n",
      "Requirement already satisfied: cycler in c:\\users\\use\\anaconda3\\lib\\site-packages (from dtale) (0.11.0)\n",
      "Requirement already satisfied: openpyxl in c:\\users\\use\\anaconda3\\lib\\site-packages (from dtale) (3.0.9)\n",
      "Requirement already satisfied: networkx in c:\\users\\use\\anaconda3\\lib\\site-packages (from dtale) (2.7.1)\n",
      "Requirement already satisfied: xlrd in c:\\users\\use\\anaconda3\\lib\\site-packages (from dtale) (2.0.1)\n",
      "Collecting missingno<=0.4.2\n",
      "  Downloading missingno-0.4.2-py3-none-any.whl (9.7 kB)\n",
      "Collecting dash-core-components==2.0.0\n",
      "  Downloading dash_core_components-2.0.0-py3-none-any.whl (3.8 kB)\n",
      "Collecting dash-html-components==2.0.0\n",
      "  Downloading dash_html_components-2.0.0-py3-none-any.whl (4.1 kB)\n",
      "Collecting dash-table==5.0.0\n",
      "  Downloading dash_table-5.0.0-py3-none-any.whl (3.9 kB)\n",
      "Requirement already satisfied: Jinja2>=2.10.1 in c:\\users\\use\\anaconda3\\lib\\site-packages (from Flask->dtale) (2.11.3)\n",
      "Requirement already satisfied: click>=5.1 in c:\\users\\use\\anaconda3\\lib\\site-packages (from Flask->dtale) (8.0.4)\n",
      "Requirement already satisfied: Werkzeug>=0.15 in c:\\users\\use\\anaconda3\\lib\\site-packages (from Flask->dtale) (2.0.3)\n",
      "Requirement already satisfied: colorama in c:\\users\\use\\anaconda3\\lib\\site-packages (from click>=5.1->Flask->dtale) (0.4.4)\n",
      "Requirement already satisfied: MarkupSafe>=0.23 in c:\\users\\use\\anaconda3\\lib\\site-packages (from Jinja2>=2.10.1->Flask->dtale) (2.0.1)\n",
      "Requirement already satisfied: tenacity>=6.2.0 in c:\\users\\use\\anaconda3\\lib\\site-packages (from plotly>=5.0.0->dtale) (8.0.1)\n",
      "Collecting brotli\n",
      "  Downloading Brotli-1.0.9-cp39-cp39-win_amd64.whl (383 kB)\n",
      "Requirement already satisfied: pyparsing>=2.2.1 in c:\\users\\use\\anaconda3\\lib\\site-packages (from matplotlib->dtale) (3.0.4)\n",
      "Requirement already satisfied: packaging>=20.0 in c:\\users\\use\\anaconda3\\lib\\site-packages (from matplotlib->dtale) (21.3)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in c:\\users\\use\\anaconda3\\lib\\site-packages (from matplotlib->dtale) (2.8.2)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in c:\\users\\use\\anaconda3\\lib\\site-packages (from matplotlib->dtale) (4.25.0)\n",
      "Requirement already satisfied: pillow>=6.2.0 in c:\\users\\use\\anaconda3\\lib\\site-packages (from matplotlib->dtale) (9.0.1)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in c:\\users\\use\\anaconda3\\lib\\site-packages (from matplotlib->dtale) (1.3.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in c:\\users\\use\\anaconda3\\lib\\site-packages (from pandas->dtale) (2021.3)\n",
      "Requirement already satisfied: charset-normalizer~=2.0.0 in c:\\users\\use\\anaconda3\\lib\\site-packages (from requests->dtale) (2.0.4)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in c:\\users\\use\\anaconda3\\lib\\site-packages (from requests->dtale) (1.26.9)\n",
      "Requirement already satisfied: idna<4,>=2.5 in c:\\users\\use\\anaconda3\\lib\\site-packages (from requests->dtale) (3.3)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in c:\\users\\use\\anaconda3\\lib\\site-packages (from scikit-learn->dtale) (2.2.0)\n",
      "Requirement already satisfied: joblib>=0.11 in c:\\users\\use\\anaconda3\\lib\\site-packages (from scikit-learn->dtale) (1.1.0)\n",
      "Requirement already satisfied: patsy>=0.5.2 in c:\\users\\use\\anaconda3\\lib\\site-packages (from statsmodels->dtale) (0.5.2)\n",
      "Building wheels for collected packages: dash-colorscales, dash-daq\n",
      "  Building wheel for dash-colorscales (setup.py): started\n",
      "  Building wheel for dash-colorscales (setup.py): finished with status 'done'\n",
      "  Created wheel for dash-colorscales: filename=dash_colorscales-0.0.4-py3-none-any.whl size=62582 sha256=7c0adb6e6a34bd939c718d852c11ad6b5edc412d518193e40e9fd551ec33a405\n",
      "  Stored in directory: c:\\users\\use\\appdata\\local\\pip\\cache\\wheels\\ec\\32\\ec\\dc8a1aa820182edf280a5e26c76fc9493ae29e196735708de2\n",
      "  Building wheel for dash-daq (setup.py): started\n",
      "  Building wheel for dash-daq (setup.py): finished with status 'done'\n",
      "  Created wheel for dash-daq: filename=dash_daq-0.5.0-py3-none-any.whl size=669714 sha256=c17af7c57f0b013be306910e7709d0aaabd1cdd5c27d220ffc3559a6e253c001\n",
      "  Stored in directory: c:\\users\\use\\appdata\\local\\pip\\cache\\wheels\\53\\73\\ea\\8a422aab5cab848d88653ed3087c1fa05553c3fbd67fb6ab75\n",
      "Successfully built dash-colorscales dash-daq\n",
      "Installing collected packages: brotli, Flask-Compress, dash-table, dash-html-components, dash-core-components, dash, strsimpy, squarify, missingno, lz4, kaleido, flask-ngrok, dash-daq, dash-colorscales, dash-bootstrap-components, dtale\n",
      "  Attempting uninstall: missingno\n",
      "    Found existing installation: missingno 0.5.1\n",
      "    Uninstalling missingno-0.5.1:\n",
      "      Successfully uninstalled missingno-0.5.1\n",
      "Successfully installed Flask-Compress-1.13 brotli-1.0.9 dash-2.6.2 dash-bootstrap-components-1.2.1 dash-colorscales-0.0.4 dash-core-components-2.0.0 dash-daq-0.5.0 dash-html-components-2.0.0 dash-table-5.0.0 dtale-2.8.1 flask-ngrok-0.0.25 kaleido-0.2.1 lz4-4.0.2 missingno-0.4.2 squarify-0.4.3 strsimpy-0.2.1\n"
     ]
    }
   ],
   "source": [
    "!pip install dtale"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cd12f1ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import dtale"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f81f0d0b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gender</th>\n",
       "      <th>race/ethnicity</th>\n",
       "      <th>parental level of education</th>\n",
       "      <th>lunch</th>\n",
       "      <th>test preparation course</th>\n",
       "      <th>math score</th>\n",
       "      <th>reading score</th>\n",
       "      <th>writing score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>female</td>\n",
       "      <td>group B</td>\n",
       "      <td>bachelor's degree</td>\n",
       "      <td>standard</td>\n",
       "      <td>none</td>\n",
       "      <td>72</td>\n",
       "      <td>72</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>female</td>\n",
       "      <td>group C</td>\n",
       "      <td>some college</td>\n",
       "      <td>standard</td>\n",
       "      <td>completed</td>\n",
       "      <td>69</td>\n",
       "      <td>90</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>female</td>\n",
       "      <td>group B</td>\n",
       "      <td>master's degree</td>\n",
       "      <td>standard</td>\n",
       "      <td>none</td>\n",
       "      <td>90</td>\n",
       "      <td>95</td>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>male</td>\n",
       "      <td>group A</td>\n",
       "      <td>associate's degree</td>\n",
       "      <td>free/reduced</td>\n",
       "      <td>none</td>\n",
       "      <td>47</td>\n",
       "      <td>57</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>male</td>\n",
       "      <td>group C</td>\n",
       "      <td>some college</td>\n",
       "      <td>standard</td>\n",
       "      <td>none</td>\n",
       "      <td>76</td>\n",
       "      <td>78</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gender race/ethnicity parental level of education         lunch  \\\n",
       "0  female        group B           bachelor's degree      standard   \n",
       "1  female        group C                some college      standard   \n",
       "2  female        group B             master's degree      standard   \n",
       "3    male        group A          associate's degree  free/reduced   \n",
       "4    male        group C                some college      standard   \n",
       "\n",
       "  test preparation course  math score  reading score  writing score  \n",
       "0                    none          72             72             74  \n",
       "1               completed          69             90             88  \n",
       "2                    none          90             95             93  \n",
       "3                    none          47             57             44  \n",
       "4                    none          76             78             75  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('stud.csv')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "db86a440",
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "df = sns.load_dataset(\"titanic\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fe2a9c0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\use\\anaconda3\\lib\\site-packages\\dtale\\views.py:725: FutureWarning:\n",
      "\n",
      "['embarked', 'class', 'deck', 'embark_town'] did not aggregate successfully. If any error is raised this will raise in a future version of pandas. Drop these columns/ops to avoid this warning.\n",
      "\n",
      "C:\\Users\\use\\anaconda3\\lib\\site-packages\\dtale\\views.py:1929: FutureWarning:\n",
      "\n",
      "The default value of regex will change from True to False in a future version.\n",
      "\n",
      "C:\\Users\\use\\anaconda3\\lib\\site-packages\\dtale\\views.py:1929: FutureWarning:\n",
      "\n",
      "The default value of regex will change from True to False in a future version.\n",
      "\n",
      "C:\\Users\\use\\anaconda3\\lib\\site-packages\\dtale\\views.py:1929: FutureWarning:\n",
      "\n",
      "The default value of regex will change from True to False in a future version.\n",
      "\n",
      "C:\\Users\\use\\anaconda3\\lib\\site-packages\\dtale\\views.py:1929: FutureWarning:\n",
      "\n",
      "The default value of regex will change from True to False in a future version.\n",
      "\n",
      "C:\\Users\\use\\anaconda3\\lib\\site-packages\\dtale\\views.py:1929: FutureWarning:\n",
      "\n",
      "The default value of regex will change from True to False in a future version.\n",
      "\n",
      "C:\\Users\\use\\anaconda3\\lib\\site-packages\\dtale\\views.py:1929: FutureWarning:\n",
      "\n",
      "The default value of regex will change from True to False in a future version.\n",
      "\n",
      "C:\\Users\\use\\anaconda3\\lib\\site-packages\\dtale\\views.py:1929: FutureWarning:\n",
      "\n",
      "The default value of regex will change from True to False in a future version.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception occurred while processing request: Could not convert value of 'y' into a numeric type.\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\plotly\\express\\_core.py\", line 343, in make_trace_kwargs\n",
      "    y = y.astype(np.float64)\n",
      "ValueError: could not convert string to float: 'Cherbourg'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\dtale\\views.py\", line 120, in _handle_exceptions\n",
      "    return func(*args, **kwargs)\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\dtale\\views.py\", line 2708, in get_column_analysis\n",
      "    return jsonify(**analysis.build())\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\dtale\\column_analysis.py\", line 137, in build\n",
      "    return_data, code = self.analysis.build(self)\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\dtale\\column_analysis.py\", line 621, in build\n",
      "    trend_line = px.scatter(x=qq_x, y=qq_y, trendline=\"ols\").data[1]\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\plotly\\express\\_chart_types.py\", line 66, in scatter\n",
      "    return make_figure(args=locals(), constructor=go.Scatter)\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\plotly\\express\\_core.py\", line 2123, in make_figure\n",
      "    patch, fit_results = make_trace_kwargs(\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\plotly\\express\\_core.py\", line 345, in make_trace_kwargs\n",
      "    raise ValueError(\n",
      "ValueError: Could not convert value of 'y' into a numeric type.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-10-09 18:30:36,276 - ERROR    - Exception occurred while processing request: Could not convert value of 'y' into a numeric type.\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\plotly\\express\\_core.py\", line 343, in make_trace_kwargs\n",
      "    y = y.astype(np.float64)\n",
      "ValueError: could not convert string to float: 'Cherbourg'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\dtale\\views.py\", line 120, in _handle_exceptions\n",
      "    return func(*args, **kwargs)\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\dtale\\views.py\", line 2708, in get_column_analysis\n",
      "    return jsonify(**analysis.build())\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\dtale\\column_analysis.py\", line 137, in build\n",
      "    return_data, code = self.analysis.build(self)\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\dtale\\column_analysis.py\", line 621, in build\n",
      "    trend_line = px.scatter(x=qq_x, y=qq_y, trendline=\"ols\").data[1]\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\plotly\\express\\_chart_types.py\", line 66, in scatter\n",
      "    return make_figure(args=locals(), constructor=go.Scatter)\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\plotly\\express\\_core.py\", line 2123, in make_figure\n",
      "    patch, fit_results = make_trace_kwargs(\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\plotly\\express\\_core.py\", line 345, in make_trace_kwargs\n",
      "    raise ValueError(\n",
      "ValueError: Could not convert value of 'y' into a numeric type.\n",
      "<__array_function__ internals>:5: RuntimeWarning:\n",
      "\n",
      "Converting input from bool to <class 'numpy.uint8'> for compatibility.\n",
      "\n",
      "<__array_function__ internals>:5: RuntimeWarning:\n",
      "\n",
      "Converting input from bool to <class 'numpy.uint8'> for compatibility.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception occurred while processing request: Must provide 'func' or tuples of '(column, aggfunc).\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\dtale\\views.py\", line 120, in _handle_exceptions\n",
      "    return func(*args, **kwargs)\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\dtale\\views.py\", line 1750, in reshape_data\n",
      "    instance = startup(data=builder.reshape(), ignore_duplicate=True)\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\dtale\\data_reshapers.py\", line 48, in reshape\n",
      "    return self.builder.reshape(data)\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\dtale\\data_reshapers.py\", line 62, in reshape\n",
      "    pivot_data = pd.pivot_table(\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\pandas\\core\\reshape\\pivot.py\", line 95, in pivot_table\n",
      "    table = __internal_pivot_table(\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\pandas\\core\\reshape\\pivot.py\", line 165, in __internal_pivot_table\n",
      "    agged = grouped.agg(aggfunc)\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\pandas\\core\\groupby\\generic.py\", line 865, in aggregate\n",
      "    relabeling, func, columns, order = reconstruct_func(func, **kwargs)\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\pandas\\core\\apply.py\", line 1260, in reconstruct_func\n",
      "    raise TypeError(\"Must provide 'func' or tuples of '(column, aggfunc).\")\n",
      "TypeError: Must provide 'func' or tuples of '(column, aggfunc).\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-10-09 18:33:51,512 - ERROR    - Exception occurred while processing request: Must provide 'func' or tuples of '(column, aggfunc).\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\dtale\\views.py\", line 120, in _handle_exceptions\n",
      "    return func(*args, **kwargs)\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\dtale\\views.py\", line 1750, in reshape_data\n",
      "    instance = startup(data=builder.reshape(), ignore_duplicate=True)\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\dtale\\data_reshapers.py\", line 48, in reshape\n",
      "    return self.builder.reshape(data)\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\dtale\\data_reshapers.py\", line 62, in reshape\n",
      "    pivot_data = pd.pivot_table(\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\pandas\\core\\reshape\\pivot.py\", line 95, in pivot_table\n",
      "    table = __internal_pivot_table(\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\pandas\\core\\reshape\\pivot.py\", line 165, in __internal_pivot_table\n",
      "    agged = grouped.agg(aggfunc)\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\pandas\\core\\groupby\\generic.py\", line 865, in aggregate\n",
      "    relabeling, func, columns, order = reconstruct_func(func, **kwargs)\n",
      "  File \"C:\\Users\\use\\anaconda3\\lib\\site-packages\\pandas\\core\\apply.py\", line 1260, in reconstruct_func\n",
      "    raise TypeError(\"Must provide 'func' or tuples of '(column, aggfunc).\")\n",
      "TypeError: Must provide 'func' or tuples of '(column, aggfunc).\n"
     ]
    }
   ],
   "source": [
    "d = dtale.show(df)\n",
    "d.open_browser()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e367bfa",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
